presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/ValueStore.java

```diff
@@ -0,0 +1,187 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation.histogram;
+
+import com.facebook.presto.array.IntBigArray;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
+import org.openjdk.jol.info.ClassLayout;
+
+import static com.facebook.presto.operator.aggregation.histogram.GroupedTypedHistogram.ProbeType.SUM_OF_SQUARE;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.calculateMaxFill;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.computeBucketCount;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextBucketId;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextSumOfCount;
+import static com.facebook.presto.operator.aggregation.histogram.HashUtil.nextSumOfSquares;
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INSUFFICIENT_RESOURCES;
+import static com.google.common.base.Preconditions.checkState;
+import static it.unimi.dsi.fastutil.HashCommon.murmurHash3;
+
+/**
+ * helper class for {@link GroupedTypedHistogram}
+ * May be used for other cases that need a simple hash for values
+ * <p>
+ * sort of a FlyWeightStore for values--will return unique number for a value. If it exists, you'll get the same number. Class map Value -> number
+ * <p>
+ * Note it assumes you're storing # -> Value (Type, Block, position, or the result of the ) somewhere else
+ */
+public class ValueStore
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final float MAX_FILL_RATIO = 0.5f;
+    private static final GroupedTypedHistogram.ProbeType PROBE_TYPE = SUM_OF_SQUARE;
+    private static final int EMPTY_BUCKET = -1;
+    private final BlockBuilder values;
+    private int rehashCount;
+
+    private int mask;
+    private int bucketCount;
+    private IntBigArray buckets;
+    private IntBigArray valueHashList;
```

Comment:
's' vs "List" because it's not an actually list?

Reply:
plus i see that's more consistent

Reply:
done

Reply:
> 's' vs "List" because it's not an actually list?

Mostly because there's no information content in saying it's a "list". The plural form conveys the same information with fewer words. It's easier to read the code.

Reply:
yea, true. got it